src/DefaultBuilder/src/WebApplicationBuilder.cs

```diff
@@ -189,48 +187,18 @@ private void ConfigureApplication(WebHostBuilderContext context, IApplicationBui
 
             if (context.HostingEnvironment.IsDevelopment())
             {
+                // TODO: add test for this
                 app.UseDeveloperExceptionPage();
             }
 
-            var implicitRouting = false;
-
-            // The endpoints were already added on the outside
-            if (_builtApplication.DataSources.Count > 0)
-            {
-                // The user did not register the routing middleware so wrap the entire
-                // destination pipeline in UseRouting() and UseEndpoints(), essentially:
-                // destination.UseRouting()
-                // destination.Run(source)
-                // destination.UseEndpoints()
-
-                // Copy endpoints to the IEndpointRouteBuilder created by an explicit call to UseRouting() if possible.
-                var targetRouteBuilder = GetEndpointRouteBuilder(_builtApplication);
-
-                if (targetRouteBuilder is null)
-                {
-                    // The app defined endpoints without calling UseRouting() explicitly, so call UseRouting() implicitly.
-                    app.UseRouting();
-
-                    // An implicitly created IEndpointRouteBuilder was addeded to app.Properties by the UseRouting() call above.
-                    targetRouteBuilder = GetEndpointRouteBuilder(app)!;
-                    implicitRouting = true;
-                }
-
-                // Copy the endpoints to the explicitly or implicitly created IEndopintRouteBuilder.
-                foreach (var ds in _builtApplication.DataSources)
-                {
-                    targetRouteBuilder.DataSources.Add(ds);
-                }
+            // Wrap the entire destination pipeline in UseRouting() and UseEndpoints(), essentially:
+            // destination.UseRouting()
+            // destination.Run(source)
+            // destination.UseEndpoints()
 
-                // UseEndpoints consumes the DataSources immediately to populate CompositeEndpointDataSource via RouteOptions,
-                // so it must be called after we copy the endpoints.
-                if (!implicitRouting)
-                {
-                    // UseRouting() was called explicitely, but we may still need to call UseEndpoints() implicitely at
-                    // the end of the pipeline.
-                    _builtApplication.UseEndpoints(_ => { });
-                }
-            }
+            // Set the route builder so we preserve the routing information that may have been set already
```

Comment:
```suggestion
            // Set the route builder so that UseRouting will use the WebApplication as the IEndpointRouteBuilder for route matching
```